### DISCUSSION TITLE
Xcode 15 beta: GRDB-related duplicated class warnings

### DISCUSSION THREAD

**Discussion by notcome** [groue/GRDB.swift#1396] [Q&A] [answered]
I am playing around with Xcode 15 and surprised to see warning messages like this:

```
objc[59263]: Class _TtC4GRDB25ValueObservationScheduler is implemented in both /Library/Developer/CoreSimulator/Volumes/iOS_21A5248u/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 17.0.simruntime/Contents/Resources/RuntimeRoot/System/Library/PrivateFrameworks/GRDB.framework/GRDB (0x1c0be5490) and /Users/xxx/Library/Developer/CoreSimulator/Devices/77039DFC-6093-4B01-A3F6-4113835E7E0F/data/Containers/Bundle/Application/4BA438C9-3327-44DD-ADC8-B54B45B98742/XXX.app/XXX (0x104325598). One of the two will be used. Which one is undefined.
```

Does this mean Apple use GRDB internally, despite just launched SwiftData? If so, congratulations!

I have two questions though:

- Is there anyway to fix this on our side?
- Is there any side effect that we should worry about? Is it possible that GRDB shipped with the system contains a bug and the code there was used over my app's embedded code?

**Comment by groue** [comment, marked as answer]
Hello @notcome,

Yes, GRDB is included in a few private SDK frameworks 😊 

I suggest reporting the problem with https://feedbackassistant.apple.com/. After all, the private SDK frameworks are not supposed to be our business in the first place. I'm not aware of any workaround on our side (which does not mean that no such workaround exist).

> Is there any side effect that we should worry about? 

According to the warning, it looks like there is a risk that mismatching classes, from different GRDB versions, will be used together, with unclear consequences (presumably bad ones). I don't know if the risk is real, or not: maybe it's better to ask Apple on https://developer.apple.com/forums/.

**Reply by notcome** [reply]
Thanks! Will definitely file a feedback.

**Reply by groue** [reply]
Thank you! I asked on the developer forums: https://developer.apple.com/forums/thread/732192

**Comment by marcomasser** [comment]
FWIW, I also filed a feedback about this with a link back to this discussion. FB12399514.

**Comment by mallman** [comment]
Whoa! Talk about a compliment! Apple themselves are using GRDB??? Hot stuff. When did this happen? Did they discuss this with you? Do you work for Apple now? Spill. @groue

**Reply by groue** [reply]
🤣 No, I don't work for Apple, but Apple does use third-party libraries :-) And I'm pretty proud indeed 😉 It happened in the previous OS release, if I remember well. The signs are pretty indirect: issues like this one, or SDK symbols extractions.

**Comment by groue** [comment]
It is solved in beta 2 :-)

**Reply by marcomasser** [reply]
I can confirm that! 🥳

**Reply by marcomasser** [reply]
Correction: It is resolved on iOS 17 Beta 2, but the warnings do still show up in visionOS 1 Beta 2:
> objc[75486]: Class _TtC4GRDB18SQLStatementCursor is implemented in both /Library/Developer/CoreSimulator/Volumes/xrOS_21N5165g/Library/Developer/CoreSimulator/Profiles/Runtimes/xrOS 1.0.simruntime/Contents/Resources/RuntimeRoot/System/Library/PrivateFrameworks/GRDB.framework/GRDB (0x1f4620110) and /Users/marco/Library/Developer/Xcode/DerivedData/SomeApp-dnggwvtgmyfoeeebxuqhnuvtlueh/Build/Products/Debug-xrsimulator/PackageFrameworks/GRDB_5DC4DB053_PackageProduct.framework/GRDB_5DC4DB053_PackageProduct (0x10559c4e8). One of the two will be used. Which one is undefined.

**Reply by groue** [reply]
So we can expect that the next beta will remove the problem :-)
